.cursor/rules/data-encoding.mdc

```diff
@@ -0,0 +1,292 @@
+---
+description: 
+globs: 
+alwaysApply: true
+---
+ ---
+description: Guidelines for encoding and decoding data in WAVS components
+globs: components/**/*
+---
+# Data Encoding and Decoding Guidelines
+
+Guidelines for handling data encoding and decoding in WAVS components.
+
+## Input Decoding
+
+1. Trigger Event Decoding:
+```rust
+fn decode_trigger_event(trigger_data: TriggerData) -> Result<(u64, Vec<u8>, Destination)> {
+    match trigger_data {
+        // Handle Ethereum contract events
+        TriggerData::EthContractEvent(TriggerDataEthContractEvent { log, .. }) => {
+            // Decode event using generated types
+            let event: solidity::NewTrigger = decode_event_log_data!(log)?;
+            let trigger_info = solidity::TriggerInfo::abi_decode(&event._triggerInfo, false)?;
+            Ok((trigger_info.triggerId, trigger_info.data.to_vec(), Destination::Ethereum))
+        }
+        // Handle raw CLI input
+        TriggerData::Raw(data) => Ok((0, data.clone(), Destination::CliOutput)),
+        _ => Err(anyhow::anyhow!("Unsupported trigger data type")),
+    }
+}
+```
+
+2. Raw Input Processing:
+```rust
+// Convert bytes to string
+let input = std::str::from_utf8(&req)
+    .map_err(|e| e.to_string())?;
+
+// Parse first character as hex digit
+let value = input
+    .chars()
+    .next()
+    .ok_or("Empty input")?
+    .to_digit(16)
+    .ok_or("Invalid hex digit")? as u64;
+```
+
+## Output Encoding
+
+1. Basic Output Encoding:
+```rust
+fn encode_trigger_output(trigger_id: u64, output: impl AsRef<[u8]>) -> Vec<u8> {
+    solidity::DataWithId {
```

Comment:
Will note that this type is an example... you can encode any solidity type. Not sure if it needs to be broken out into a separate function. Maybe we could more easily illustrate this by explicitly showing these types with the `sol!` macro? Like how I did here: https://github.com/Lay3rLabs/hats-protocol-avs-example/blob/main/components/hats-eligibility/src/lib.rs

It should be clear that it's a pattern and that components are free to encode or decode any types they want.

Reply:
solidity::DataWithId is from the sol! macro -- this is found in the example Solidity contract. This type is generic here but really it is forced into structure by wavs. Fixed on 0.4 with the new "WasmResult" name 

if DataWithId is not provided then WAVS does not know how to process submission